helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -385,10 +389,19 @@ public static int getStateCount(String state, StateModelDefinition stateModelDef
           // If the desired state is the top state, but the instance cannot be transited to the
           // top state in one hop, try to keep the top state on current host or a host with a closer
           // state.
-          for (String currentStatePrioritizedInstance : sortedPreferenceList) {
-            if (!assigned.contains(currentStatePrioritizedInstance) && liveAndEnabled
-                .contains(currentStatePrioritizedInstance)) {
+          while (currentStatePrioritizedInstanceIter.hasNext()) {
+            // Note that it is safe to check the prioritized instance items only once here.
+            // Since the only possible condition when we don't use an instance in this list is that
+            // it has been assigned with some state. And this is not revertable in this method. So
+            // checking it one more time later will only waste time.
+            String currentStatePrioritizedInstance = currentStatePrioritizedInstanceIter.next();
+            if (!assigned.contains(currentStatePrioritizedInstance)) {
               proposedInstance = currentStatePrioritizedInstance;
+              // If we find a different instance for the partition placement, then we need to check
+              // the same instance again or it will not be assigned with any partitions.
+              if (!proposedInstance.equals(instance)) {
+                i--;
```

Comment:
This is a trade-off. 
I agree queue would be much simpler. And with the queue, I can do more optimization.
However, since this calculation will be run many times. Creating a queue here might prolong the rebalance time a lot. Let me see if I can profile this.

Reply:
I tried it, with some enhancement, the performance is actually slightly better.
The only downside is that we introduced many more changes to make it clearer. Please review the new change.

Reply:
One thing to note here, offer does not work. Adding the item to the back will dramatically change the preference list order, that is not what we want.